diff --git a/gdbstub/gdbstub.c b/gdbstub/gdbstub.c
index b1def7e..a02b29a 100644
--- a/gdbstub/gdbstub.c
+++ b/gdbstub/gdbstub.c
@@ -1438,7 +1438,7 @@ static void handle_v_kill(GArray *params, void *user_ctx)
     gdb_put_packet("OK");
     error_report("QEMU: Terminated via GDBstub");
     gdb_exit(0);
-    gdb_qemu_exit(0);
+    exit(0);
 }
 
 static const GdbCmdParseEntry gdb_v_commands_table[] = {
@@ -2091,7 +2091,7 @@ static int gdb_handle_packet(const char *line_buf)
         /* Kill the target */
         error_report("QEMU: Terminated via GDBstub");
         gdb_exit(0);
-        gdb_qemu_exit(0);
+        exit(0);
         break;
     case 'D':
         {
